File: examples/nodejs/generate_invoice.js
// Rubrol Node.js Client Example
// Requires Node 18+ (built-in fetch)
const fs = require('fs');

const RUBROL_URL = 'http://localhost:8080/v1/render';

async function generateInvoice() {
  const payload = {
    template: 'b2b_invoice',
    data: {
      invoice_number: 'INV-2026-NODE01',
      issued_date: '2026-09-17',
      due_date: '2026-10-17',
      currency: 'USD',
      currency_symbol: '$',
      vendor: {
        name: 'Vercel Edge Solutions',
        tax_id: 'US-8819203',
        address: '440 N Barranca Ave',
        city: 'Covina, CA 91723',
        email: 'billing@vercel.com'
      },
      customer: {
        name: 'Supabase Data Labs',
        tax_id: 'US-2291048',
        address: '970 Toa Payoh North',
        city: 'Singapore 318992',
        email: 'accounts@supabase.io'
      },
      line_items: [
        {
          description: 'High-Volume Document Sidecar Worker Pod',
          qty: 2,
          unit_price: 245.00
        }
      ]
    },
    format: 'pdf',
    pdf_standard: 'a-2b'
  };

  console.log(`Sending payload to ${RUBROL_URL}...`);
  const start = Date.now();

  const response = await fetch(RUBROL_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });

  if (!response.ok) {
    throw new Error(`HTTP Error: ${response.status} ${response.statusText}`);
  }

  const renderTimeMs = response.headers.get('X-Render-Time-Ms');
  const buffer = Buffer.from(await response.arrayBuffer());

  const filename = 'nodejs_invoice.pdf';
  fs.writeFileSync(filename, buffer);

  console.log(`Generated ${filename} (${buffer.length} bytes)`);
  console.log(`Engine Render Latency: ${renderTimeMs}ms | Round-trip: ${Date.now() - start}ms`);
}

generateInvoice().catch(console.error);

File: examples/python/generate_invoice.py
# -*- coding: utf-8 -*-
"""
Rubrol Python Client Example
Renders an invoice via the local Rubrol sidecar and writes the resulting PDF.
"""
import urllib.request
import json
import time

RUBROL_URL = "http://localhost:8080/v1/render"

payload = {
    "template": "b2b_invoice",
    "data": {
        "invoice_number": "INV-2026-PY01",
        "issued_date": "2026-09-17",
        "due_date": "2026-10-17",
        "currency": "USD",
        "currency_symbol": "$",
        "vendor": {
            "name": "Acme SaaS Cloud Inc.",
            "tax_id": "US-9918204",
            "address": "100 Pine Street, Suite 2400",
            "city": "San Francisco, CA 94111",
            "email": "billing@acme.dev"
        },
        "customer": {
            "name": "Global Logistics Corp",
            "tax_id": "US-5541092",
            "address": "450 Lexington Ave",
            "city": "New York, NY 10017",
            "email": "ap@globallogistics.com"
        },
        "line_items": [
            {
                "description": "Rubrol High-Throughput Sidecar Cluster (Annual)",
                "qty": 1,
                "unit_price": 490.00
            },
            {
                "description": "Zero-Copy Factur-X & PDF/A-2b Engine Integration",
                "qty": 1,
                "unit_price": 350.00
            }
        ]
    },
    "format": "pdf",
    "pdf_standard": "a-2b"
}

print(f"Sending render request to {RUBROL_URL}...")
start = time.perf_counter()

req = urllib.request.Request(
    RUBROL_URL,
    data=json.dumps(payload).encode("utf-8"),
    headers={"Content-Type": "application/json"}
)

with urllib.request.urlopen(req) as resp:
    pdf_bytes = resp.read()
    render_time = resp.headers.get("X-Render-Time-Ms", "N/A")

output_file = "python_invoice.pdf"
with open(output_file, "wb") as f:
    f.write(pdf_bytes)

elapsed_ms = (time.perf_counter() - start) * 1000
print(f"Success! {len(pdf_bytes):,} bytes written to {output_file}")
print(f"Sidecar Compile Time: {render_time}ms | Total HTTP Latency: {elapsed_ms:.2f}ms")
